### PR TITLE
[modbus] fix len in 0x2B func code + add PDU category detection

### PR DESCRIPTION
**Pull request by smainand**
Hello,

this patch fixes a mistake in the 0x2B function code and incorporates the functionality developed by @thmsaurel to interpret the function code category.

@thmsaurel, could you validate all the modifications you proposed are covered ?
If this PR is accepted, it replaces the #123 PR.

Regards
Sébastien

### CODE REVIEW

**Comment by thmsaurel** [comment]
Hi,
Seems great to me. This code is used inside the PR #374 (OPC DA integration by @G4RUF). Do you plan to also integrate this code inside scapy ?
Regards

**Comment by smainand** [comment]
Hello,

yes, I just started playing with opc-da but I intend to do so once I would have made more tests on it and 
 checked with @G4RUF.

Regards
Seb.

File: scapy/contrib/modbus.py [hunk 1]
```diff
@@ -14,10 +14,7 @@
 # You should have received a copy of the GNU General Public License
 # along with Scapy. If not, see <http://www.gnu.org/licenses/>.
 
-# scapy.contrib.description = ModBus Protocol
-# scapy.contrib.status = loads
```
Comment (gpotter2) [inline scapy/contrib/modbus.py]: Hey, I've spent some time to make hose comments right :/
Could you leave them ?

Comment (smainand) [inline scapy/contrib/modbus.py]: Hello, 
my bad, it is now fixed. Thanks for your review.
Seb.

File: scapy/contrib/modbus.py [hunk 2]
```diff
@@ -574,10 +538,166 @@ class ModbusPDU2B0EReadDeviceIdentificationError(Packet):
...
+
+
+class ModbusPDUReservedFunctionCodeRequest(Packet):
+    name = "Reserved Function Code Request"
+    fields_desc = [
+            ByteEnumField("funcCode", 0x00, _reserved_funccode_request),
+            StrFixedLenField('payload', '', 255), ]
+
+    def extract_padding(self, s):
+        return "", None
```
Comment (gpotter2) [inline scapy/contrib/modbus.py]: Could you use b"" instead ? (We're porting scapy to python 3)

File: scapy/contrib/modbus.py [hunk 3]
```diff
@@ -574,10 +538,166 @@ class ModbusPDU2B0EReadDeviceIdentificationError(Packet):
...
+
+
+class ModbusPDUReservedFunctionCodeError(Packet):
+    name = "Reserved Function Code Error"
+    fields_desc = [
+            ByteEnumField("funcCode", 0x00, _reserved_funccode_error),
+            StrFixedLenField('payload', '', 255), ]
+
+    def extract_padding(self, s):
+        return "", None
```
Comment (gpotter2) [inline scapy/contrib/modbus.py]: Same

File: scapy/contrib/modbus.py [hunk 4]
```diff
@@ -574,10 +538,166 @@ class ModbusPDU2B0EReadDeviceIdentificationError(Packet):
...
+class ModbusPDUUserDefinedFunctionCodeRequest(Packet):
+    name = "User-Defined Function Code Request"
+    fields_desc = [
+            ModbusByteEnumField(
+                "funcCode", 0x00, _userdefined_funccode_request,
+                "Unknown user-defined request function Code"),
+            StrFixedLenField('payload', '', 255), ]
+
+    def extract_padding(self, s):
+        return "", None
```
Comment (gpotter2) [inline scapy/contrib/modbus.py]: Idem

File: scapy/contrib/modbus.py [hunk 5]
```diff
@@ -574,10 +538,166 @@ class ModbusPDU2B0EReadDeviceIdentificationError(Packet):
...
+class ModbusPDUUserDefinedFunctionCodeError(Packet):
+    name = "User-Defined Function Code Error"
+    fields_desc = [
+            ModbusByteEnumField(
+                "funcCode", 0x00, _userdefined_funccode_error,
+                "Unknown user-defined error function Code"),
+            StrFixedLenField('payload', '', 255), ]
+
+    def extract_padding(self, s):
+        return "", None
```
Comment (gpotter2) [inline scapy/contrib/modbus.py]: Same

File: scapy/contrib/modbus.py [hunk 6]
```diff
@@ -574,10 +538,166 @@ class ModbusPDU2B0EReadDeviceIdentificationError(Packet):
...
+
+
+class ModbusPDUReservedFunctionCodeResponse(Packet):
+    name = "Reserved Function Code Response"
+    fields_desc = [
+            ByteEnumField("funcCode", 0x00, _reserved_funccode_response),
+            StrFixedLenField('payload', '', 255), ]
+
+    def extract_padding(self, s):
+        return "", None
```
Comment (gpotter2) [inline scapy/contrib/modbus.py]: Same

File: scapy/contrib/modbus.py [hunk 7]
```diff
@@ -574,10 +538,166 @@ class ModbusPDU2B0EReadDeviceIdentificationError(Packet):
...
+    name = "User-Defined Function Code Response"
+    fields_desc = [
+            ModbusByteEnumField(
+                "funcCode", 0x00, _userdefined_funccode_response,
+                "Unknown user-defined response function Code"),
+            StrFixedLenField('payload', '', 255), ]
+
+
+    def extract_padding(self, s):
+        return "", None
```
Comment (gpotter2) [inline scapy/contrib/modbus.py]: Same

**Comment by gpotter2** [comment]
Looks definatly ok otherwise.

Calling @guedou for review

File: scapy/contrib/modbus.py (hunk 5 above)
Comment (smainand) [inline scapy/contrib/modbus.py]: Hello,
here is the fix (16 occurrences)
Seb.

File: scapy/contrib/modbus.py [hunk 8]
```diff
@@ -17,7 +17,7 @@
 # scapy.contrib.description = ModBus Protocol
 # scapy.contrib.status = loads
 
-# Copyright (C) 2016 Arthur Gervais, Ken LE PRADO, Sébastien Mainand
+# Copyright (C) 2016 Arthur Gervais, Ken LE PRADO, Sébastien Mainand, Thomas Aurel
```
Comment (guedou) [inline scapy/contrib/modbus.py]: s/2016/2017/

Comment (smainand) [inline scapy/contrib/modbus.py]: Hello, 
it is done
Seb.

**Pull request merged**